tools/shared.py

```diff
@@ -2652,21 +2652,44 @@ def safe_copy(src, dst):
   if dst == '/dev/null': return
   shutil.copyfile(src, dst)
 
-def clang_preprocess(filename):
-  # TODO: REMOVE HACK AND PASS PREPROCESSOR FLAGS TO CLANG.
-  return subprocess.check_output([CLANG_CC, '-DFETCH_DEBUG=1', '-E', '-P', '-C', '-x', 'c', filename])
-
-def read_and_preprocess(filename):
-  f = open(filename, 'r').read()
-  pos = 0
-  include_pattern = re.compile('^#include\s*["<](.*)[">]\s?$', re.MULTILINE)
-  while(1):
-    m = include_pattern.search(f, pos)
-    if not m:
-      return f
-    included_file = open(os.path.join(os.path.dirname(filename), m.groups(0)[0]), 'r').read()
+def find_lines_with_preprocessor_directives(file):
+  return filter(lambda x: '#if' in x or '#elif' in x, open(file, 'r').readlines())
+
+def run_c_preprocessor_on_file(src, dst):
+  # Run LLVM's C preprocessor on the given file, expanding #includes and variables found in src/setting.js.
```

Comment:
> how about doing this in python

I did not want to create a second implementation of the partial C preprocessor in `src/parseTools.js`. Using LLVM to preprocess can be considered Correct out of the box, so that won't need to care about adding tests, especially tests to check that python preprocessor and `src/parseTools.js` would agree. Writing a preprocessor is hard (#5457, #5458, [[3]](https://github.com/kripken/emscripten/blob/master/src/compiler.js#L167)). I took a pass long time ago at fixing these types of issues with a more complete preprocessor in JS (#1969), but its complexity was too high that we did not want to land it.

I don't think I'm able to implement a preprocessor in Python that I would feel good about, that would either have less complexity than the approach at #1969 had, or if settling for less supported features, not have various similar kinds of limits that people eventually run into with the current JS preprocessor. That is why I felt really glad to notice I can just reuse LLVM's preprocessor here, albeit with a slight change of behavior. Getting the ability to do `#define`, `#if A || B` and `#if !A && B` for "free" are big improvements, so I'd strongly like to use LLVM to be able to get to use those.

Before implementing the whitelist heuristic, I wrote a blacklisting heuristic, with manually listed identifiers that should not get expanded, which only contained `TOTAL_MEMORY` for the purposes of `pthread-main.js` and `Fetch.js`. That felt a bit more dirty than this.

Some alternatives to be able to drop the whitelisting mechanism:
  - expand the `src/settings.js` defines to code with an alias. For example, we could give all defines there a `S_` prefix, so that `TOTAL_MEMORY` would become `-DS_TOTAL_MEMORY=1` for the preprocessor.
  - Don't expand `src/settings.js` defines to code automatically, but have a directive in `src/Settings.js`, with something like `exportToPreprocessor('TOTAL_MEMORY', (optional)'S_TOTAL_MEMORY');` which would tell this machinery to take the `TOTAL_MEMORY` variable and preprocess with it, possibly using the specified optional alias.
  - Expand `src/settings.js` defines automatically, but add a `hideFromPreprocessor('TOTAL_MEMORY');` directive or similar to allow removing some fields, or a `exportToPreprocessor('TOTAL_MEMORY', (optional)'S_TOTAL_MEMORY');` as well to allow routing the defines under other names.

Overall, running C preprocessor on our .js files is the best way we have to offer as efficient as possible form of "don't pay for what you don't use" type of dead code elimination, to be able to get to tiny compiled applications that a lot of people are looking for. Having the "full" LLVM preprocessor available to do that would be quite helpful here. Is there some variant above that you think would work better than the current whitelist mechanism in this PR?